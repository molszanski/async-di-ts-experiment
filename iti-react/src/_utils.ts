export type UnPromisify<T> = T extends Promise<infer U> ? U : T

export type GetContainerFormat<ProviderFunction extends (...args: any) => any> =
  UnPromisify<ReturnType<ProviderFunction>>

export function addGetter(object, key, fn: any) {
  Object.defineProperty(object, key, {
    get() {
      return fn()
    },
    enumerable: true,
  })
}
