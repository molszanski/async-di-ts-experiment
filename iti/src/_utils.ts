export type UnPromisify<T> = T extends Promise<infer U> ? U : T

export type GetContainerFormat<ProviderFunction extends (...args: any) => any> =
  UnPromisify<ReturnType<ProviderFunction>>

export function addGetter(object, key, fn: any) {
  Object.defineProperty(object, key, {
    get() {
      return fn()
    },
    enumerable: true,
  })
}

export type Assign4<OldContext extends object, NewContext extends object> = {
  [Token in keyof ({
    [K in keyof OldContext]: OldContext[K]
  } & {
    [K in keyof NewContext]: NewContext[K]
  })]: Token extends keyof NewContext
    ? NewContext[Token]
    : Token extends keyof OldContext
    ? OldContext[Token]
    : never
}

export type Prettify<T> = T extends infer U ? { [K in keyof U]: U[K] } : never

export type UnpackFunction<T> = T extends (...args: any) => infer U ? U : T
export type UnpackObject<T> = {
  [K in keyof T]: UnpackFunction<T[K]>
}

export type UnpromisifyObject<T> = {
  [K in keyof T]: UnPromisify<T[K]>
}
// keep
// type AssignAndUnpackObjects<O1 extends {}, O2 extends {}> = UnpromisifyObject<
//   UnpackObject<Assign4<O1, O2>>
// >

export type FullyUnpackObject<T extends {}> = UnpromisifyObject<UnpackObject<T>>

export type KeysOrCb<Context extends {}> =
  | Array<keyof Context>
  | ((t: { [K in keyof Context]: K }) => Array<keyof Context>)

export type MyRecord<O extends {}, T> = {
  [K in keyof O]: T
}
export type ContextGetter<Context extends {}> = {
  [CK in keyof Context]: UnpackFunction<Context[CK]>
}

export type UnpackFunctionReturn<T> = T extends (arg: T) => infer U ? U : T
export type ContextGetterWithCache<Context extends {}> = {
  [CK in keyof Context]: UnpackFunctionReturn<Context[CK]>
}

export function _intersectionKeys(
  needle: { [key: string]: any },
  haystack: { [key: string]: any },
) {
  let haystackKeys = Object.keys(haystack)
  let duplicates = haystackKeys.filter((x) => x in needle)
  if (duplicates.length === 0) {
    return undefined
  }
  return duplicates.join("', '")
}
