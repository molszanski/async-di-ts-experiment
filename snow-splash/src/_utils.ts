export type UnPromisify<T> = T extends Promise<infer U> ? U : T

export type GetContainerFormat<ProviderFunction extends (...args: any) => any> =
  UnPromisify<ReturnType<ProviderFunction>>

export function addGetter(object, key, fn: any) {
  Object.defineProperty(object, key, {
    get() {
      return fn()
    },
    enumerable: true,
  })
}

export type Assign4<OldContext extends object, NewContext extends object> = {
  [Token in keyof ({
    [K in keyof OldContext]: OldContext[K]
  } & {
    [K in keyof NewContext]: NewContext[K]
  })]: Token extends keyof NewContext
    ? NewContext[Token]
    : Token extends keyof OldContext
    ? OldContext[Token]
    : never
}
