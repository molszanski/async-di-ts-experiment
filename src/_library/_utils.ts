export const wait = (w: number) => new Promise((r) => setTimeout(r, w))

export type UnPromisify<T> = T extends Promise<infer U> ? U : T

export type GetContainerFormat<ProviderFunction extends (...args: any) => any> =
  UnPromisify<ReturnType<ProviderFunction>>
