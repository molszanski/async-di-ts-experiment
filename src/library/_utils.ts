export const wait = (w: number) => new Promise((r) => setTimeout(r, w))
