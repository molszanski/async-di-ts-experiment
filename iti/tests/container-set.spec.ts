import { getMainMockAppContainer } from "./mocks/_mock-app-container"
import { wait } from "./_utils"

it("should get two containers that are async", async () => {
  const cont = getMainMockAppContainer()
  let containerSet = await cont.getContainerSet(["aCont", "bCont"])

  expect(containerSet).toHaveProperty("aCont")
  expect(containerSet).toHaveProperty("bCont")
  expect(containerSet.bCont.b2).toMatchObject({ a1: {} })

  expect(containerSet).toMatchSnapshot(containerSet)
})

it("should subscribe to container set change", (cb) => {
  ;(async () => {
    const cont = getMainMockAppContainer()
    let containerSet = await cont.getContainerSet(["aCont", "bCont", "cCont"])

    expect(containerSet).toHaveProperty("aCont")
    expect(containerSet).toHaveProperty("bCont")
    expect(containerSet.bCont.b2).toMatchObject({ a1: {} })
    expect(containerSet.cCont.c2.size).toBe(5)

    containerSet.cCont.upgradeCContainer()
    cont.subscribeToContainerSet(
      ["aCont", "bCont", "cCont"],
      (err, containerSet) => {
        expect(containerSet.cCont.c2.size).toBe(10)
      },
    )
    await cont.get("cCont")
    await wait(10)
    cb()
  })()
})

it("should get container set via a new API", (cb) => {
  ;(async () => {
    const cont = getMainMockAppContainer()
    let containerSet = await cont.getContainerSet((c) => [c.aCont, c.bCont])

    expect(containerSet).toHaveProperty("aCont")
    expect(containerSet).toHaveProperty("bCont")
    expect(containerSet.bCont.b2).toMatchObject({ a1: {} })
    expect(containerSet).toMatchSnapshot(containerSet)

    cb()
  })()
})

it("should subscribe to container set change via a new APi", (cb) => {
  ;(async () => {
    const cont = getMainMockAppContainer()
    let containerSet = await cont.getContainerSet((c) => [c.aCont, c.cCont])
    expect(containerSet).toHaveProperty("aCont")

    cont.subscribeToContainerSet(
      (c) => {
        return [c.aCont, c.cCont]
      },
      (err, containerSet) => {
        expect(containerSet.cCont.c2.size).toBe(10)
      },
    )
    containerSet.cCont.upgradeCContainer()
    await wait(10)
    cb()
  })()
})

it("should subscribe to container set change via a old APi", (cb) => {
  ;(async () => {
    const cont = getMainMockAppContainer()
    let containerSet = await cont.getContainerSet(["aCont", "cCont"])
    expect(containerSet).toHaveProperty("aCont")

    cont.subscribeToContainerSet(
      (c) => {
        return [c.aCont, c.cCont]
      },
      (err, containerSet) => {
        expect(containerSet.cCont.c2.size).toBe(10)
      },
    )

    containerSet.cCont.upgradeCContainer()
    await wait(10)
    cb()
  })()
})

it("should be able to unsubscribe from container set change", (cb) => {
  ;(async () => {
    const cont = getMainMockAppContainer()
    let containerSet = await cont.getContainerSet((c) => [c.aCont, c.cCont])

    const fn = jest.fn()

    const unsub = cont.subscribeToContainerSet(
      (c) => [c.cCont],
      () => {
        unsub()
        fn()

        cont.subscribeToContainerSet(
          (c) => [c.cCont],
          () => {
            // TODO: well this should be an error probably
            expect(fn).toHaveBeenCalledTimes(2)
          },
        )
        containerSet.cCont.upgradeCContainer()
        wait(10).then(() => cb())
      },
    )
    containerSet.cCont.upgradeCContainer()
  })()
})
