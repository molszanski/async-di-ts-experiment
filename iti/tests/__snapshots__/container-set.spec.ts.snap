// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get container set via a new API 1`] = `
Object {
  "aCont": Object {
    "a1": Object {},
    "a2": Object {
      "a1": Object {},
    },
    "a3": Object {
      "a1": Object {},
      "a2": Object {
        "a1": Object {},
      },
    },
  },
  "bCont": Object {
    "b1": Object {
      "a2": Object {
        "a1": Object {},
      },
    },
    "b2": Object {
      "a1": Object {},
    },
  },
}
`;

exports[`should get two containers that are async 1`] = `
Object {
  "aCont": Object {
    "a1": Object {},
    "a2": Object {
      "a1": Object {},
    },
    "a3": Object {
      "a1": Object {},
      "a2": Object {
        "a1": Object {},
      },
    },
  },
  "bCont": Object {
    "b1": Object {
      "a2": Object {
        "a1": Object {},
      },
    },
    "b2": Object {
      "a1": Object {},
    },
  },
}
`;
